[JENKINS-48463] Update jenkins core to use  XML v1.1

mikecirioli opened the pull request:
See: [JENKINS-48463](https://issues.jenkins-ci.org/browse/JENKINS-48463)

Currently jenkins creates (and supports) xml v1.0 documents, which can cause issues if a user has any forbidden (ie. control-XX) characters as part of a job or other configuration file due to the usage of non-xstream classes for parsing XML in various plugins and components.  XStream itself does not care about XML versions, but the underlying XML Pull Parser that was being used only supports XML v1.0

This PR updates the classes that create files using XStream to emit a xml v1.1 prologue, and to use the KXml2Driver for XML parsing (the mxpp driver that was being used did not support XML v1.1).

Unit tests have been updated to support XML v1.1 as well.

For more information, please see [Valid_characters_in_XML](https://en.wikipedia.org/wiki/Valid_characters_in_XML#Non-restricted_characters)

### Proposed changelog entries

* Upgrade Jenkins core from XML 1.0 to 1.1 (configuration files prologues are updated)
    * Change XStream driver to KXml2Driver, needed for xml v1.1 support
    * Update XStream classes to emit xml v1.1 prologues

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@jtnord @svanoort @batmat @varmenise 

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

mikecirioli committed:
[oss-2590] updated dependencies to support move to xml 1.1

mikecirioli committed:
[JENKINS-48463] - update jenkins core to use xml v1.1
  - switched to KxmlDriver (needed for xml 1.1 support)
  - updated to create xml v1.1 headers
  - updated test to use KXml2Driver explicitly
  - updated unit tests to support xml v1.1

mikecirioli committed:
[JENKINS-48463] update hard coded references of XPP3Driver to KXml2Driver in order to support XMLv1.1

oleg-nenashev commented on core/pom.xml:
```diff
@@ -446,11 +451,6 @@ THE SOFTWARE.
       <version>${spring.version}</version>
     </dependency>
-    <dependency>
-      <groupId>xpp3</groupId>
```
Are there any transient dependencies on this library in plugins?
There is a risk of breaking the binary compatibility of the core.

I also see there are usages in the XStream fork: https://github.com/search?q=org%3Ajenkinsci+Xpp3Driver&type=Code . Version is the same: https://github.com/jenkinsci/xstream/blob/d3a5bd43086303169f4737758bf89be2e5584a71/pom.xml#L716.

Due to the XStream dependency, probably it's safe to remove it from here

batmat replied:
we can do a `mvn dependency:list` before and after this PR to assess this @mikecirioli.

oleg-nenashev reviewed (changes requested):
IMHO the PR needs compatibility investigation and explicit automatic migration tests:

* Jenkins can start from the JENKINS_HOME with XML 1.0 
* When the configuration is saved, files get updated to XML v 1.1

mikecirioli commented:
I’ve verified the migration use case anecdotally, but I agree that an explicit test case for this is worth writing.

I will investigate the xpp3 dependency more this weekend.

Thanks for the feedback.
> On Dec 8, 2017, at 6:38 PM, Oleg Nenashev <notifications@github.com> wrote:
> 
> @oleg-nenashev requested changes on this pull request.
> 
> IMHO the PR needs compatibility investigation and explicit automatic migration tests:
> 
> Jenkins can start from the JENKINS_HOME with XML 1.0
> When the configuration is saved, files get updated to XML v 1.1
> In core/pom.xml:
> 
> > @@ -446,11 +451,6 @@ THE SOFTWARE.
>        <artifactId>spring-aop</artifactId>
>        <version>${spring.version}</version>
>      </dependency>
> -    <dependency>
> -      <groupId>xpp3</groupId>
> Are there any transient dependencies on this library in plugins?
> There is a risk of breaking the binary compatibility of the core.
> 
> I also see there are usages in the XStream fork: https://github.com/search?q=org%3Ajenkinsci+Xpp3Driver&type=Code . Version is the same: https://github.com/jenkinsci/xstream/blob/d3a5bd43086303169f4737758bf89be2e5584a71/pom.xml#L716.
> 
> Due to the XStream dependency, probably it's safe to remove it from here
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

oleg-nenashev commented:
Thanks @mikecirioli! 
Please also note that the commits will need to be rebased/squashed in order to replace custom issues IDs with Jenkins ones.

jtnord commented on core/pom.xml:
```diff
@@ -240,6 +240,11 @@ THE SOFTWARE.
+    <dependency>
+        <groupId>net.sf.kxml</groupId>
+        <artifactId>kxml2</artifactId>
+        <version>2.3.0</version>
```
what where the reasons for not using the stock JDK SAX parser which is XML 1.1 compliant?

mikecirioli replied:
I was having issues getting Xstream to use the correct parser if i recall correctly.  I've got a bit better understanding of how things work now, and will go back and see if that is still the case.

mikecirioli replied:
@jtnord - after quite a bit of investigation, I was able to use the stock JDK sax parser, but this introduced a number of new issues.  
- The stock parser wants to write its own <?xml ...> header, and when using the HierarchicalStreamDriver (which is what XStream2 uses), you are not given the option to change it (and its defaulted to <?xml version='1.0'>
- Switching to use an XMLStreamWriter gives you the ability to set both the character encoding and xml version.  However, this then requires a fairly significant amount of change in our XStream implementation.

I think that for now the best choice is to use the KXml2 driver.  It is more performant than the Xpp3 driver, and if we see value in moving to a stock JDK implementation in the future, we can tackle that as a separate issue.

Side Note: The original implementation was a bit inconsistent in which drivers were being used.  I've cleaned that up to ensure that all of the XML operations are now using the same driver.

jtnord commented on core/src/test/java/jenkins/model/RunIdMigratorTest.java:
```diff
@@ -83,13 +83,13 @@
     @Test public void legacy() throws Exception {
         assumeFalse("Symlinks don't work well on Windows", Functions.isWindows());
-        write("2014-01-02_03-04-05/build.xml", "<?xml version='1.0' encoding='UTF-8'?>\n<run>\n  <stuff>ok</stuff>\n  <number>99</number>\n  <otherstuff>ok</otherstuff>\n</run>");
+        write("2014-01-02_03-04-05/build.xml", "<?xml version='1.1' encoding='UTF-8'?>\n<run>\n  <stuff>ok</stuff>\n  <number>99</number>\n  <otherstuff>ok</otherstuff>\n</run>");
```
so I think these changes are wrong - they are testing you can migrate from an old build layout to the new layout - as such the files would have been written with the old XStream utilities and be `xml 1.0`

mikecirioli replied:
but when they get migrated they will be written out as xml 1.1.  I'll take a closer look and see what i can do about that.

mikecirioli replied:
After looking into what the migrator does, there are no XStream operations involved, so xml version will not be changed.  Reverting these changes

jtnord reviewed (changes requested):
I think the run migration tests should be left as xml 1.0.

Can you say why we are using the kxml2 parser vs a stock JDK SAX/DOM parser (they should all be xml 1.1 compliant).

Can we have a test using an old config file for Jenkins that shows we can still read a 1.0 written config file.

mikecirioli commented:
@jtnord I am writing tests for verifying that legacy (xml 1.0) configs are still properly consumed.

batmat commented on core/src/test/java/jenkins/model/RunIdMigratorTest.java:
```diff
@@ -83,13 +83,13 @@
+        assertEquals("{99={build.xml='<?xml version='1.1' encoding='UTF-8'?>\n<run>\n  <stuff>ok</stuff>\n  <id>2014-01-02_03-04-05</id>\n  <timestamp>1388649845000</timestamp>\n  <otherstuff>ok</otherstuff>\n</run>'}, lastFailedBuild=→-1, lastSuccessfulBuild=→99, legacyIds='2014-01-02_03-04-05 99\n'}", summarize());
         assertEquals(99, migrator.findNumber("2014-01-02_03-04-05"));
         migrator = new RunIdMigrator();
         assertFalse(migrator.migrate(dir, null));
```
I agree with @jtnord IIUC think you are right to do that replacement, but only after that `migrator.migrate` call, which is expected to migrate things from old 1.0 to new.

There can't be a Jenkins version using xml 1.1, AND using old IDs.

mikecirioli replied:
Stepping through the migrator code, its just doing basic File operations (no xstream) so in this case it doesn't matter about the XML version.  I will revert them back to '1.0'

batmat reviewed (changes requested):
RunIdMigrationTest ought to be adjusted (1.0 must stay as the starting point). 

The rest LGTM.

Thanks for moving this forward!

oleg-nenashev commented:
#TooManyLabelsException. @mikecirioli are you still working on it?

mikecirioli commented:
I am Oleg, hoping to wrap things up this weekend.

-mike


> On Jan 6, 2018, at 11:48 AM, Oleg Nenashev <notifications@github.com> wrote:
> 
> #TooManyLabelsException. @mikecirioli <https://github.com/mikecirioli> are you still working on it?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/jenkinsci/jenkins/pull/3185#issuecomment-355759271>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AHwjT1BcKMk8V8mVzbYJOsgEBi_a5gF1ks5tH6PMgaJpZM4Q7nbg>.
>

mikecirioli committed:
[JENKINS-48463] refactored XStream2 to default to KXml2Driver by default.
Changed all explicit references of the hierarchical driver to use a static
convinience method XStream2.getDefaultDriver() to ensure all XML operations
are using the same driver.

mikecirioli committed:
[JENKINS-48463] Added additional unit tests for verifying xml v1.1 compatibility
  - verify can still read xml 1.0
  - verify can read xml 1.1
  - verify can read xml 1.1 with 'special' characters
  - verify that jenkins can start with a config.xml having
    <?xml version='1.1'> and special characters

mikecirioli commented:
@oleg-nenashev @batmat @jtnord  - i believe i've addressed the comments and changes requested.  Please review the last couple of commits so i can move this forward.

mikecirioli committed:
[JENKINS-48463] Added a unit test that validates jenkins xml v1.0 configs
are silently migrated to xml v1.1 as they are persisted

daniel-beck commented:
The people who complain about job config history text diffs are going to hate this.

mikecirioli commented:
@daniel-beck  - any thoughts on how to deal with that?  I just realized that the @localData for the last unit test i added technically should have caused the JenkinsRule to blow up (its xml v1.0 with special chars) - so it seems that the KXmlDriver is more tolerant of such things than the preview Xpp3Driver was.  This would imply that we might be able to keep the XML version at 1.0 while still allowing the special chars to exist in the config files.

I'm not 100% thrilled if that is the case, but as a compromise it might the right choice.

mikecirioli committed:
[JENKINS-48463] It appears that the KXml2Driver is more tolerant of
technically illegal xml (ie. it doesn't complain if an Xml v1.0 file
contains special characters).  Added a unit test that proves this, as well
as fixing silentlyMigrateConfigsTest so that it starts with technically
valid Xml 1.0 content

mikecirioli commented:
@daniel-beck @oleg-nenashev @jtnord FYI:

I just added another unit test to validate the original issue that spawned this work (ie. XML v1.0 with special characters should cause a SAXParserException to be thrown).  As i suspected, this test actually fails when using the KXml2Driver, as it appears to be tolerant of this technically illegal XML content.

I will do an actual manual verification that XML 1.0 + Special Chars + KXml2Driver doesn't make jenkins unhappy later this morning, but this does indeed seem to imply that we don't technically have to change the XML version header.  TBD if this is a good thing or not (it seems bad that we could be creating invalid XML, even if in practice it doesn't cause anyone issues).

daniel-beck commented:
@mikecirioli Nothing necessarily needs fixing, just pointing it out.

mikecirioli commented:
@daniel-beck in that case, I will mark the new test i added as @Ignore for now, with a comment as to why (ie. KXml is tolerant).

mikecirioli committed:
[JENKINS-48463] - Ignoring test xml1_0_withSpecialCharsShouldFail due to
KXml2Driver being tolerant of control characters which should be illegal
in an XML v1.0 content.  This test should be revisted if we switch to
another XML driver

svanoort reviewed (approved):
Seems OK to me once the merge conflict is resolved?

mikecirioli committed:
Merge branch 'master' into OSS-2590

oleg-nenashev reviewed (dismissed):
Test failures seem to be related: 'java.lang.NoClassDefFoundError: org/xmlpull/mxp1/MXParser' . 🐛 until tests are green. 

It definitely comes from the dependency changes

batmat commented:
Some tests failures
![image](https://user-images.githubusercontent.com/223853/35283210-e26d4394-0057-11e8-8836-0d33f9d6d60c.png)

mikecirioli commented:
yes, it looks like i missed a hard coded reference to the Xpp3Driver, thats been fixed and I am running all the tests locally one more time to be sure before i push that change.

mikecirioli committed:
[JENKINS-48463] - fixed reference to deprecated Xpp3Driver and replaced with
new default driver

mikecirioli commented:
@oleg-nenashev  - tests are passing now

batmat commented on core/src/main/java/hudson/util/XStream2.java:
```diff
@@ -97,7 +98,18 @@
     private MapperInjectionPoint mapperInjectionPoint;
 
+    /**
+     * Convinience method so we only have to change the driver in one place
```
Convenience typo

batmat commented on test/src/test/java/hudson/XMLFileTest.java:
```diff
@@ -0,0 +1,45 @@
+        // verify that we did indeed load our test config.xml
+        assertThat(j.jenkins.getLabelString(), is("I am a label"));
+        //verify that the persisted top level config.xml is v1.1
+        File configFile = new File(j.jenkins.getRootPath().getRemote() + File.separator + "config.xml");
```
NIT: `File` has a constructor with two params for this kind of usage: 

```java
new File(j.jenkins.getRootDir(), "config.xml");
```

should be equivalent FWIW

batmat reviewed (approved):
:bee:

mikecirioli committed:
[JENKINS-48463] Erring on the side of caution and incuding xpp3 libs back
in, just in case some plugin might be depending on them

daniel-beck commented:
Not a big fan of additional core dependencies that get exposed to plugins, our mess is big enough ☹️

mikecirioli commented:
@daniel-beck  -  any suggestions on what i can do to minimize that?  Adding the KXml2 driver is unavoidable, and from what i can determine, both it and the Xpp driver are provided by our XStream fork don't think this should make the situation worse (even if it doesn't make it better).   Its possible i am overlooking something obvious to others though.

jtnord commented on core/pom.xml:
```diff
@@ -246,6 +246,16 @@ THE SOFTWARE.
+    <dependency>
+      <groupId>xpp3</groupId>
+      <artifactId>xpp3</artifactId>
+      <version>1.1.4c</version>
```
should this have a comment this is unused but for backwards compatibility?  Otherwise I'm not sure what this is here for.

batmat replied:
+1, would be nice to have a statement about it.

jtnord commented on core/src/main/java/hudson/XmlFile.java:
```diff
@@ -352,13 +351,14 @@ private void attempt() throws Eureka {
     private static final SAXParserFactory JAXP = SAXParserFactory.newInstance();
 
-    private static final Xpp3Driver DEFAULT_DRIVER = new Xpp3Driver();
+    private static final HierarchicalStreamDriver DEFAULT_DRIVER = XStream2.getDefaultDriver();
```
Do we not want to be specific about the default driver being `kXML`?  in case a future xStream changes it and causes some issues?  (I guess tests would pick this up?)

jtnord commented on core/src/main/java/hudson/util/XStream2.java:
```diff
@@ -97,7 +98,18 @@
+     * @return a new instance of the HierarchicalStreamDriver we want to use
+     */
+    public static HierarchicalStreamDriver getDefaultDriver() {
+        return new KXml2Driver();
```
Are drivers not ThreadSafe?  Can we not reuse them?

mikecirioli replied:
http://x-stream.github.io/faq.html#Scalability

Yes, i will address this in a future PR

jtnord commented on pom.xml:
```diff
@@ -133,7 +133,7 @@ THE SOFTWARE.
         <groupId>org.apache.ant</groupId>
         <artifactId>ant</artifactId>
-        <version>1.8.4</version>
+        <version>1.9.2</version>
```
???
will this cause issues with the "ant" plugin?

jtnord commented on test/src/test/java/hudson/XMLFileTest.java:
```diff
@@ -0,0 +1,45 @@
+        File configFile = new File(j.jenkins.getRootPath().getRemote() + File.separator + "config.xml");
+        assertThat(configFile.exists(), is(true));
+        BufferedReader config = new BufferedReader(new FileReader(configFile));
+        assertThat(config.readLine(), is("<?xml version='1.1' encoding='UTF-8'?>"));
```
a failing test will leak handles which will cause tidyup to fail as the file is open
use a try-with-resources

jtnord reviewed (approved):
LGTM - unit test could be tweaked to not be nasty if it failed.  re-using the driver can be a future enhancement if the driver is thread safe

mikecirioli committed:
[JENKINS-48463] added try-with-resources to XmlFileTest to ensure proper
file handle cleanup in the event of failure

Fixed small nits regarding typo's and code cleanup
Added comment explaining why Xpp3 is still included as a dep

batmat commented on core/pom.xml:
```diff
@@ -246,6 +246,20 @@ THE SOFTWARE.
+      <groupId>xpp3</groupId>
+      <artifactId>xpp3</artifactId>
+      <version>1.1.4c</version>
+    </dependency>
```
NIT: I would have put this back at the original place, so that this does *not* show up as a change in this PR. (note: do not spend doing this, just a thought for next time).

mikecirioli replied:
ack will keep that in mind in the future

batmat reviewed (approved):
:bee:

batmat commented:
@mikecirioli could/did you build the Jenkins WAR locally and a dry run to see if nothing wrong immediately surfaces?

batmat commented:
Just did it, nothing obviously wrong. Jenkins does start up correctly, I was able to pass the wizard, install suggested plugins (checking a random file did indeed contains a prologue with XML 1.1 now in use).

mikecirioli commented:
ack, i've built and run it locally as well without any obvious issues

jtnord commented:
@reviewbybees done

batmat commented:
This change has been reviewed, and seems to be ready for merge. I will merge it later this week, or early next one if nobody objects in the meantime. Thanks everyone! Kudos @mikecirioli !

oleg-nenashev commented:
@batmat are you still going to merge it?

batmat commented:
@oleg-nenashev done. Adjusted the changelog a bit.

mikecirioli commented:
Thanks Baptiste!
-mike


> On Feb 4, 2018, at 10:49 AM, Baptiste Mathus <notifications@github.com> wrote:
> 
> @oleg-nenashev <https://github.com/oleg-nenashev> done. Adjusted the changelog a bit.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/jenkinsci/jenkins/pull/3185#issuecomment-362897859>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AHwjTwT4bz-IJNy0JzUSFz6eQC9dXwkXks5tRYtWgaJpZM4Q7nbg>.
>
